core/templates/dev/head/components/VersionDiffVisualizationDirective.js

```diff
@@ -0,0 +1,301 @@
+          $scope.diffGraphNodeColors[nodeId] = COLOR_DELETED;
+          _stateTypeUsed[NODE_TYPE_DELETED] = true;
+        } else if (nodeStateProperty === STATE_PROPERTY_CHANGED) {
+          diffGraphNodes[nodeId] = nodesData[nodeId].originalStateName;
+          $scope.diffGraphNodeColors[nodeId] = COLOR_CHANGED;
+          if (nodesData[nodeId].originalStateName !=
+              nodesData[nodeId].newestStateName) {
+            $scope.diffGraphSecondaryLabels[nodeId] = '(was: ' +
+              nodesData[nodeId].originalStateName + ')';
+            diffGraphNodes[nodeId] = nodesData[nodeId].newestStateName;
+            _stateTypeUsed[NODE_TYPE_CHANGED_RENAMED] = true;
+          } else {
+            _stateTypeUsed[NODE_TYPE_CHANGED] = true;
+          }
+        } else if (nodeStateProperty === STATE_PROPERTY_UNCHANGED) {
+          diffGraphNodes[nodeId] = nodesData[nodeId].originalStateName;
+          $scope.diffGraphNodeColors[nodeId] = COLOR_UNCHANGED;
+          if (nodesData[nodeId].originalStateName !=
+              nodesData[nodeId].newestStateName) {
+            $scope.diffGraphSecondaryLabels[nodeId] = '(was: ' +
+              nodesData[nodeId].originalStateName + ')';
+            diffGraphNodes[nodeId] = nodesData[nodeId].newestStateName;
+            $scope.diffGraphNodeColors[nodeId] = COLOR_RENAMED_UNCHANGED;
+            _stateTypeUsed[NODE_TYPE_RENAMED] = true;
+          } else {
+            _stateTypeUsed[NODE_TYPE_UNCHANGED] = true;
+          }
+        } else {
+          throw new Error('Invalid state property.');
+        }
+      }
+
+      $scope.v1InitStateId = $scope.getDiffData().v1InitStateId;
+
+      $scope.diffGraphData = {
+        nodes: diffGraphNodes,
+        links: $scope.getDiffData().links,
+        initStateId: $scope.getDiffData().v2InitStateId,
+        finalStateIds: $scope.getDiffData().finalStateIds
+      };
+
+      // Generate the legend graph
+      $scope.legendGraph = {
+        nodes: {},
+        links: []
+      };
+      var _lastUsedStateType = null;
+      for (var stateProperty in _stateTypeUsed) {
+        if (_stateTypeUsed[stateProperty]) {
+          $scope.legendGraph.nodes[stateProperty] = stateProperty;
+          if (_lastUsedStateType) {
+            $scope.legendGraph.links.push({
+              source: _lastUsedStateType,
+              target: stateProperty,
+              linkProperty: 'hidden'
+            });
+          }
+          _lastUsedStateType = stateProperty;
+          if (!$scope.legendGraph.hasOwnProperty('initStateId')) {
+            $scope.legendGraph.initStateId = stateProperty;
+          }
+        }
+      }
+      $scope.legendGraph.finalStateIds = [_lastUsedStateType];
+
+      // Functions to show modal of history diff of a state
+      // stateId is the unique ID assigned to a state during calculation of
+      // state graph.
+      $scope.onClickStateInHistoryGraph = function(stateId) {
```

Comment:
Maybe call this DiffGraph instead of HistoryGraph? (Since it's going to be used more generally.)

Same for the docstring above.